### ISSUE TITLE
Affine bijector does not register training variable

### ISSUE THREAD

**Issue by andompesta** [tensorflow/probability#683]
I'm trying to replicate the Neormalization flow demo provided by [Eric Jang](https://blog.evjang.com/2018/01/nf1.html) using tensorflow 2 and tensorflow probability.

I'm focusing on the first part of the demo. However, my flow does not train because it does not has any trained variable registered so there is no gradient computed. I'm not sure if the way I initialize the affine transformation is wrong or if I should use a different approach.
You can find my code here:
```python
import numpy as np
import tensorflow as tf
import tensorflow_probability as tfp
from tensorflow_probability import distributions as tfd
from tensorflow_probability import bijectors as tfb
import matplotlib.pyplot as plt


BATCH_SIZE = 512
D, R = 2, 2
NUL_LAYER = 6

def target_distribution(batch_size=BATCH_SIZE):
    """
    generate sample from the target distribution
    p(x1, x2) = N(x1| mu=1/4 * x2^2, sigma=1) * N(x2| mu=0, sigma=4)
    :param batch_size: batch size of the sample
    :return:
    """
    x2_dis = tfd.Normal(loc=0., scale=4.)
    print(x2_dis)
    x2_sample = x2_dis.sample(batch_size)
    x1_dist = tfd.Normal(loc=0.25 * tf.square(x2_sample),
                    scale=tf.ones(batch_size, dtype=tf.float32))
    print(x1_dist)
    x1_sample = x1_dist.sample()
    x = tf.stack([x1_sample, x2_sample], axis=1)
    return x

class LeakyReLU(tfb.Bijector):
    def __init__(self, alpha=0.5, validate_args=False, name="leaky_relu"):
        super(LeakyReLU, self).__init__(forward_min_event_ndims=1, inverse_min_event_ndims=1, validate_args=validate_args, name=name)
        self.alpha = alpha

    def _forward(self, x):
        return tf.where(tf.greater_equal(x, 0), x, self.alpha * x)

    def _inverse(self, y):
        return tf.where(tf.greater_equal(y, 0), y, 1. / self.alpha * y)

    def _inverse_log_det_jacobian(self, y):
        # we assume y to be batch_size * dim
        event_dims = self.inverse_min_event_ndims
        I = tf.ones_like(y)
        J_inv = tf.where(tf.greater_equal(y, 0), I, 1.0 / self.alpha * I)       # it is expected to be a 2x2 matrix but a vector is faster
        # abs is actually redundant here, since this det Jacobian is > 0
        log_abs_J_inv = tf.math.log(tf.abs(J_inv))
        return tf.reduce_sum(log_abs_J_inv, axis=event_dims)



if __name__ == '__main__':
    var_inizializars = tf.initializers.VarianceScaling()
    target_sample = target_distribution(BATCH_SIZE)

    # plt.figure(figsize=(4, 4))
    # plt.scatter(target_smaple[:, 0], target_smaple[:, 1], s=5)
    # plt.axis('equal')
    # plt.show()

    prior = tfd.MultivariateNormalDiag(loc=tf.zeros(2),
                                       scale_diag=tf.ones(2))

    # build normalization flows

    bijectors = []
    for i in range(NUL_LAYER):
        bij = tfb.Affine(
            shift=var_inizializars([D]),
            scale_tril=tfd.fill_triangular(var_inizializars([D * (D + 1) // 2])),
            scale_perturb_factor=var_inizializars([D, R])

        )
        bijectors.append(bij)
        # add nonlinearity (activation)
        alpha = tf.abs(var_inizializars([])) + .01
        lrelu = LeakyReLU(alpha)
        bijectors.append(lrelu)

    mlp_bijectors = tfb.Chain(bijectors[:-1])
    flow = tfd.TransformedDistribution(prior,
                                       bijector=mlp_bijectors)

    NUM_STEPS = int(1e5)
    optim = tf.optimizers.Adam(lr=3e-5)
    for i in range(NUM_STEPS):
        with tf.GradientTape() as tape:
            log_prb = flow.log_prob(target_sample)
            loss = -tf.reduce_mean(log_prb)

        gradients = tape.gradient(loss, flow.trainable_variables)
        optim.apply_gradients(zip(gradients, flow.trainable_variables))
        print(i, loss.numpy())
```

**Comment by csuter** [comment]
Are you using tf-nightly and tfp-nightly packages? We haven't yet prepared a stable release including all the changes we've recently made to support instantiation of Distriubtions and Bijectors  outside of the gradient tape. Can you try with tf-nightly and tfp-nightly and let us know if you still have issues?

**Comment by andompesta** [comment]
tfp-nightlyHmm, I have installed tf-nightly and tf-nightly-gpu then I have installed tfp-nightly, but the problem persist. No variable is detected as trainable. Should I wrap the TransformedDistribution inside a keras model ?

**Comment by csuter** [comment]
Ah, apologies -- the Affine bijector was one that we didn't migrate, but deprecated instead. The "argument zoo" style was going to be hard to make work in the new style, wherein we avoid doing any "work" (computation) at instance initialization time.

The recommended replacement is to explicitly chain together `Shift` bijector and one of several `Scale` bijectors. Since you're combining scale_tril and scale_perturb_factor you'll probably want to construct a LinearOperatorLowRankUpdate instance and use `tfb.ScaleMatvecLinearOperator`. Something like this:

```
tril = tf.linalg.LinearOperatorLowerTriangular(
    tfd.fill_triangular(var_inizializars([D * (D + 1) // 2])))

lru = tf.linalg.LinearOperatorLowRankUpdate(
    tril, u=var_inizializars([D, R]))

shift=tfb.Shift(var_inizializars([D]))

# Applies from right to left: scale first, then shift
bij = tfb.Chain([shift, tfb.ScaleMatvecLineraOperator(lru)])
```

Sorry for the incomplete advice in the previous comment!

**Comment by SiegeLordEx** [comment]
See also https://github.com/tensorflow/probability/issues/448#issuecomment-555629330 for a conversion chart between the now deprecated `Affine` and the replacement bijectors.

**Comment by andompesta** [comment]
Thanks for the advice. I manage to get it to work.
One more stupid error that I found was in the variable generation:
apparently any initializer ``var_inizializars = tf.initializers.VarianceScaling()`` return an EagerTensor not a ``tf.Variable`` this was the reason why I do not have any trainable variable in my flow.
Coming from pytorch this is a big old fashion.

A solution to this problem is:
```
tensor_inizializar = tf.initializers.GlorotUniform()
var_inizalizer = lambda x: tf.Variable(tensor_inizializar(x))
```
but I'm wandering if it's better to have a special arguments that specify the initializers to returns variables instead of tensors

**Comment by csuter** [comment]
Hm, sorry for the confusion (and for not noticing it myself when looking at your code!). I think this is "working as intended" but could perhaps be documented better on the TF side? In short, these Initializers don't intrinsically have anything to do with Variables, even if they are very often used to initialize tf.Variables in the TF/NN setting, they just map shapes (and presumably random seeds?) to some set of values, which may be used as starting values for tf.Variables, or potentially something else (like maybe chain states for initializing MCMC w/ TFP).

I think your solution is nice, fwiw!
